src/api/libcellml/parser.h

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2015 University of Auckland
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef LIBCELLML_LIBCELLML_PARSER_H_
+#define LIBCELLML_LIBCELLML_PARSER_H_
+
+#include <string>
+
+#include "libcellml/model.h"
+#include "libcellml/types.h"
+#include "libcellml/units.h"
+#include "libcellml/logger.h"
+
+//! Everything in libCellML is in this namespace.
+namespace libcellml {
+
+/**
+ * @brief The Parser class.
+ *
+ * The Parser class is for representing a CellML Parser.
+ */
+class LIBCELLML_EXPORT Parser: public Logger
+{ 
+public:
+    Parser(Format format); /**< Constructor */
+    ~Parser(); /**< Destructor */
+    Parser(const Parser &rhs); /**< Copy constructor */
+    Parser(Parser &&rhs); /**< Move constructor */
+    Parser& operator=(Parser p); /**< Assignment operator */
+
+    /**
+     * @brief Create and populate a new model from a @c std::string.
+     *
+     * Creates and populates a new model pointer by parsing CellML
+     * entities and attributes from the @p input @c std::string.
+     *
+     * @param input The string to parse into a model.
+     *
+     * @return The new @c ModelPtr deserialised from the input string.
+     */
+    ModelPtr parseModel(const std::string &input);
+
+    /**
+     * @brief Updates an existing model with the attributes from a @c std::string.
+     *
+     * Updates the argument @p model with CellML entities and attributes
+     * from the argument @p input @c std::string. Any entities, objects,
+     * or attributes of with matching names in the model will be
+     * overwritten by those loaded from the @p input @c std::string.
+     *
+     * @param model The @c ModelPtr to update.
+     * @param input The string to parse into the @p model.
+     */
+    void updateModel(const ModelPtr &model, const std::string &input);
```

Comment:
Good point- this is currently set up so that `parseModel()` just creates a new model, calls `updateModel()` and returns the model. But we should probably remove `updateModel()` or at least make it private until we want to deal with updates after validation.